bip-ccv.mediawiki

```diff
@@ -0,0 +1,413 @@
+
+We consider the ''program'' of a P2TR UTXO to be composed of an x-only public key (that we call the ''naked key''), and a Merkle tree of Scripts. If there is no data committed in the UTXO, then the naked key is the internal key as defined in BIP-341.
+
+If the UTXO carries a commitment to a 32-byte hash (the ''data''), the naked key is tweaked with a hash of the data. The resulting key is the taproot internal key per BIP-341.
+
+This allows to embed a commitment to the data that can be validated during the Script execution, while staying fully compatible with taproot. Notably:
+* the committed data does not make the UTXO any larger;
+* the keypath spend is still available to any party that possesses the private key of the naked key, as long as they have knowledge of the embedded data (or at least their hash)<ref>For example, in a multi-party contract, the naked key could be aggregate key using [[bip-0327.mediawiki|MuSig2]]; the taproot keypath would therefore allow a ''cooperative'' spend, without executing any Script at all. Like for all taproot transactions, this is indeed the cheapest way of spending the UTXO - albeit not always possible in practice.</ref>;
+* if multiple Scripts are in different leaves of the taptree, only the ones that actually need to access the data have to pay a cost for it, in terms of additional witness bytes.
+
+<code>OP_CHECKCONTRACTVERIFY</code> can be applied to introspect the program and data of one of the inputs of the transaction (typically, the current input, in order to access its committed data), or one of the outputs of the transaction (in order to define its program, and possibly its committed data).
+
+=== Output amounts ===
+
+When checking the Script of one or more output with <code>OP_CHECKCONTRACTVERIFY</code>, it is usually necessary to also check that the amount of the current UTXO is correctly distributed among the outputs in the expected way. Therefore, the opcode already includes an amount semantic that covers the common use cases.
+
+There are three supported modes for the opcode when checking an output, depending on the value of the <code>mode</code> parameter:
+* ''default'': the residual amount of the current input must be preserved in the output (aggregate across the inputs that specify the output);
+* ''ignore'': the output amount is ignored.
+* ''deduct'': the amount of the checked output is subtracted from the amount of the current input (the residual amount is then available for further checks);
+
+The ''default'' logic covers the common case where an input wishes to send the full amount to a specific output.
+
+The ''deduct'' logic allows to assign portions of the input amount to one or more outputs; the residual amount, checked with a final check using the ''default'' logic, can be used to enforce that the total amount is preserved.
```

Comment:
If there are multiple UTXOs, what happens with the change? If there is only one UTXO, how do you spend the entire amount without bringing more funds to pay for fees?

Reply:
What is 'change'? The amount has to be entirely assigned to the outputs.
For applications like vaults, you can do use the _deduct_ logic with an output that has the same script as the input; most contracts are non-recursive, so they always 'move forward', and there is no meaningful definition of change.

Fees have to be paid exogenously, with a separate input (or by spending an output, if that applies). I think this is necessary in any kind of covenant, as the fee is by definition an amount that is not encumbered by the covenant.

Reply:
I may have not realized then yet that the fees are always paid exogenously.